File: sample-python/snippets/staticmethod.py
# (C) A.Voß – codebasedlearning.dev – Study Maths and Computer Science (AMI) with us @ FH Aachen – https://ami.codebasedlearning.dev

""" ---- Content ----

Teaching Focus
- `@staticmethod` – a !![#decorator] for a plain function in a class.
- What `@staticmethod` gives up, and what it buys.
- When `@classmethod` is the better answer.

Idea Codebook
- In this snippet we structured the code such that the main discussion
  points are shown and navigable inside the Codebook-Window.
- Answers to questions can be shown or hidden, same for definitions or
  concepts such as 'Decorator'.
- In fact, text is rendered, so you can also show some links
  [PEP 8 - Style Guide for Python Code](https://peps.python.org/pep-0008)
  or html <p>
  <center><img src="../docs/Logo_CBL_2024_72.png" width="50"></center>

See also
- !![#decorator]
"""

""" ---- Header ---- """

from typing import Self
from utils import print_function_header

""" ---- Discussion ---- """

class Temperature:
    """ --- `Temperature` - A class with all three kinds of functions. ---

    `Temperature` carries one value and three callables that differ only in what
    Python hands them as the first argument:

    | Kind            | First argument | Typical use                     |
    |-----------------|----------------|---------------------------------|
    | instance method | `self`         | needs the object's data         |
    | `@classmethod`  | `cls`          | alternative constructor         |
    | `@staticmethod` | nothing        | helper that belongs to the type |
    """

    ZERO_CELSIUS_IN_KELVIN = 273.15

    def __init__(self, celsius: float):
        self.celsius = celsius

    def as_fahrenheit(self) -> float:
        """ -- `as_fahrenheit` – The instance method --
        It reads `self.celsius`, so it needs an object `self`. """
        return self.celsius * 9 / 5 + 32

    @classmethod
    def from_kelvin(cls, kelvin: float) -> Self:
        """ -- .`from_kelvin` – The class method --
        It constructs an instance via `cls`.

        `cls` is the class the call was made on, so a subclass gets *its* own
        type back - which is exactly why `Temperature(...)` is not hardcoded.
        """
        return cls(kelvin - cls.ZERO_CELSIUS_IN_KELVIN)

    @staticmethod
    def is_plausible(celsius: float) -> bool:
        """ -- .`is_plausible` – The static method --
        No `self`, no `cls`: a plain function in the class's [#namespace].

        It validates a value without needing any object at all. Written as a
        module function it would work identically; written here it stays where a
        reader looks for it.
        """
        return -273.15 <= celsius <= 5000.0

@print_function_header
def calling_functions():
    """ --- `calling_functions` ---

    Note the call syntax via class or instance or both.

    Rule:
    - Does the callable need the class it was called on? -> `@classmethod`.
    - Does it need neither the object nor the class? -> `@staticmethod`.
    - Does it need the object's data? -> A plain method.

    Code:
    - `from_kelvin` needs `cls`
    - `is_plausible` needs nothing - it only reads its argument
    """

    t = Temperature(20.0)

    # call via instance `t`
    print(f" 1| as_fahrenheit()   -> {t.as_fahrenheit()}")

    # call via class `Temperature`
    print(f" 2| from_kelvin(300)  -> {Temperature.from_kelvin(300.0).celsius:.2f}")

    # call via class or instance
    print(f" 3| via class         -> {Temperature.is_plausible(20.0)}")
    print(f" 4| via instance      -> {t.is_plausible(20.0)}")


@print_function_header
def why_cls_is_important():
    """ --- `why_cls_is_important` ---

    !![What if we do not use `return cls(...)` in `from_kelvin` but
    `return Temperature(...)`?](#a-001)
    """

    class Kelvin(Temperature):
        pass

    t = Temperature.from_kelvin(300.0)
    print(f" 1| Temperature.from_kelvin(300)  -> {type(t).__name__=}")

    k = Kelvin.from_kelvin(300.0)
    print(f" 2| Kelvin.from_kelvin type  -> {type(k).__name__=}")

""" ---- Run ---- """

if __name__ == "__main__":
    """ --- `main` ---
    - !![#main-guard]
    - !![#code-style]
    - !![#pythonic-way]
    """
    calling_functions()
    why_cls_is_important()

File: sample-cpp/snippets/staticmethod.cpp
// (C) A.Voß – codebasedlearning.dev – Study Maths and Computer Science (AMI) with us @ FH Aachen – https://ami.codebasedlearning.dev

/* ---- Content ----
 *
 * Teaching Focus
 * - `static` member functions – a plain function in a class's !![#scope]
 * - What `static` gives up, and what it buys.
 * - Why C++ has no `@classmethod`, and what a !![#named-constructor] replaces.
 *
 * Idea Codebook
 * - We structured the code such that the main discussion points are shown
 *   and navigable inside the Codebook-Window.
 * - Answers to questions can be shown or hidden, same for definitions or
 *   concepts such as !![#named-constructor].
 * - Text is rendered, so you can also show some links
 *   [C++ Core Guidelines](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines)
 *   or plain html <p>
 *   <center><img src="../docs/Logo_CBL_2024_72.png" width="50"></center>
 *
 * See also
 * - !![#code-style]
 */

/* ---- Header ---- */

#include <cstdlib>
#include <iostream>

#include "utils/printing.h"
using std::cout, std::endl;

/* ---- Discussion ---- */

/* --- `Temperature` - A class with `static` functions. ---
 *
 * `Temperature` carries one value and three callables that differ only in what
 * the compiler hands them as the (hidden) first argument:
 *
 * | Kind                   | Hidden argument | Typical use                     |
 * |------------------------|-----------------|---------------------------------|
 * | member function        | `this`          | needs the object's data         |
 * | static member function | nothing         | helper that belongs to the type |
 * | [#named-constructor]   | nothing         | build an object, named clearly  |
 *
 * Python has a third kind, `@classmethod`, which receives the class itself.
 * C++ has no equivalent.
 */
class Temperature {
public:
    static constexpr double zero_celsius_in_kelvin = 273.15;

    explicit Temperature(double celsius) : celsius_{celsius} {}

    /* -- `as_fahrenheit` – The member function --
     * It reads `celsius_`, so it needs an object - the hidden `this`. <p>
     * `const` promises not to modify it: the `this` pointer is `const Temperature*` inside.
     */
    [[nodiscard]] double as_fahrenheit() const { return celsius_ * 9.0 / 5.0 + 32.0; }

    /* -- .`from_kelvin` – The static named constructor --
       A static member function that returns an object - what C++ uses where
       Python writes an alternative constructor as a `@classmethod`.

       Unlike Python's `cls`, it has no idea which class the call was written
       on.
     */
    [[nodiscard]] static Temperature from_kelvin(double kelvin) { return Temperature{kelvin - zero_celsius_in_kelvin}; }

    /* -- .`is_plausible` – The static member function --
       No `this`: an ordinary function that happens to live in the class's
       [#scope].

       It validates a value without needing any object at all. Written as a free
       function it would compile to the same thing; written here it stays where
       a reader looks for it.
     */
    [[nodiscard]] static bool is_plausible(double celsius) { return -273.15 <= celsius && celsius <= 5000.0; }

    [[nodiscard]] double celsius() const { return celsius_; }

private:
    double celsius_;
};

/* --- `calling_functions` ---
 * Note the call syntax via class or object or both. <p>
 * A static member function is reachable through the class and through an
 * object - the object is evaluated and then discarded, because there is no
 * `this` to pass.
 * - !![#background-static-via-object]
 *
 * Rule:
 * - Does the callable need the object's data? -> A member function.
 * - Does it need neither the object nor its type? -> `static`.
 * - Does it build an object and deserve a name? -> A named constructor.
 *
 * Code:
 * - `as_fahrenheit` needs `this`
 * - `is_plausible` needs nothing - it only reads its argument
 */
void calling_functions() {
    print_function_header(__func__);

    const Temperature t{20.0};

    // call via object `t`
    cout << " 1| as_fahrenheit()   -> " << t.as_fahrenheit() << endl;

    // call via class `Temperature`
    cout << " 2| from_kelvin(300)  -> " << Temperature::from_kelvin(300.0).celsius() << endl;

    // both calls valid
    cout << " 3| via class         -> " << Temperature::is_plausible(20.0) << endl;
    cout << " 4| via object        -> " << t.is_plausible(20.0) << endl;
}

// internal linkage, see below
static int file_local_calls = 0;

/* --- `three_meanings_of_static` ---
 * We have
 * - internal linkage – not visible to other [#translation-unit]s
 * - function-local – initialized once, on first use, and outlives the call
 * - class member – belongs to the class, not to an object
 */
void three_meanings_of_static() {
    print_function_header(__func__);

    // function-local
    static int local_calls = 0;
    ++local_calls;
    ++file_local_calls;

    cout << " 1| local_calls       -> " << local_calls << endl;
    cout << " 2| file_local_calls  -> " << file_local_calls << endl;

    // class member
    cout << " 3| class constant    -> " << Temperature::zero_celsius_in_kelvin << endl;

    if (local_calls < 2) {
        three_meanings_of_static();         // call again to show the difference
    }
}

/* --- `why_there_is_no_cls` ---
 * Assume we call `from_kelvin` on a derived class. <br>
 * !![What is the type?](#a-001)
 */
void why_there_is_no_cls() {
    print_function_header(__func__);

    /* -- .Inheritance without `cls` --
     * A derived class inherits static member functions, so `Kelvin::is_plausible`
     * compiles - but `from_kelvin` still constructs a `Temperature`, because
     * nothing tells it where the call was written.
     */
    struct Kelvin : Temperature {
        using Temperature::Temperature;
    };

    cout << " 1| Kelvin::is_plausible(-500)  -> " << Kelvin::is_plausible(-500.0) << endl;
    cout << " 2| Kelvin::from_kelvin(300)    -> " << Kelvin::from_kelvin(300.0).celsius() << endl;
}

/* ---- Run ---- */

int main() {
    calling_functions();
    three_meanings_of_static();
    why_there_is_no_cls();

    return EXIT_SUCCESS;
}
